modules/tftp/http_downloads.rb

```diff
@@ -0,0 +1,318 @@
+require 'uri'
+require 'securerandom'
+require 'digest'
+require 'json'
+require 'concurrent'
+
+module Proxy::TFTP
+  class HttpError < RuntimeError
+    attr_reader :code, :message
+
+    def initialize(code, message)
+      @code = code
+      @message = message
+    end
+  end
+
+  class HttpDownload
+    class Status
+      attr_accessor :last_error, :timestamp, :stopped
+
+      def initialize(now = Time.now)
+        @m = Mutex.new
+        @file_length = 0
+        @downloaded = 0
+        @timestamp = now
+        @stopped = false
+      end
+
+      def file_length
+        @m.synchronize { @file_length }
+      end
+
+      def downloaded
+        @m.synchronize { @downloaded }
+      end
+
+      def file_length=(length)
+        @m.synchronize { @file_length = length }
+      end
+
+      def update_downloaded(chunk_length)
+        @m.synchronize { @downloaded += chunk_length }
+      end
+
+      def progress
+        (fl = file_length) == 0 ? 0 : ((downloaded.to_f/fl) * 100).round(2).to_i
```

Comment:
No. This uses a mutex, you'll get a "ThreadError: deadlock" if you attempt to acquire a mutex multiple times from the same thread.  This is fine as ```file_length``` and ```downloaded``` are initialized/updated at different times and individually synchronized.